File: Data Structures & Algorithms/permutation-string/submission-9.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if len(s2) < len(s1): 
            return False 

        
        b1 = [0] * 26 
        for c in s1:
            b1[ord(c) - ord('a')] += 1 

        b2 = [0] * 26
        for i in range(len(s1)): 
            b2[ord(s2[i]) - ord('a')] += 1 

        if b1 == b2: 
            return True 

        l = 0
        for r in range(len(s1), len(s2)): 
            deleted = s2[l] 
            cur = s2[r] 
            b2[ord(deleted) - ord('a')] -= 1 
            b2[ord(cur) - ord('a')] += 1  
            l += 1
            if b1 == b2: 
                return True 

        return False

        

File: Data Structures & Algorithms/minimum-window-with-characters/submission-9.js
class Solution {
    /**
     * @param {string} s
     * @param {string} t
     * @return {string}
     */
    minWindow(s, t) {
        let need = {};
        for (const c of t) {
            need[c] = ( need[c] || 0 ) + 1
        }

        let l = 0;
        let matches = 0;
        let start = 0;
        let minLen = s.length + 1;
        let needLen = Object.keys(need).length;
        for (let r = 0; r < s.length; r ++) {
            let cur = s[r];
            if (need[cur] !== undefined) {
                need[cur] --;
                if (need[cur] === 0) {
                    matches ++
                }
            }

            while(matches === needLen) {
                if ((r - l + 1) < minLen) {
                    minLen = r - l + 1;
                    start = l
                }
                let deleted = s[l];
                if (need[deleted] !== undefined) {
                    need[deleted] ++;
                    if (need[deleted] > 0) {
                        matches --
                    }
                }
                l++
            }
        }

        return minLen === s.length + 1 ? "" : s.substring(start, start + minLen)
    }
}
